plugin-api/src/main/java/org/hyperledger/besu/plugin/services/storage/KeyValueStorage.java

```diff
@@ -61,15 +63,23 @@
   Optional<byte[]> get(byte[] key) throws StorageException;
 
   /**
-   * Performs an evaluation against each key in the store, keeping the entries that pass, removing
-   * those that fail.
+   * Returns a stream of all keys.
    *
-   * @param retainCondition predicate to evaluate each key against, unless the result is {@code
-   *     null}, both the key and associated value must be removed.
-   * @return the number of keys removed.
-   * @throws StorageException problem encountered when removing data.
+   * @return A stream of all keys in storage.
+   * @throws StorageException problem encountered during the retrieval attempt.
+   */
+  Stream<byte[]> streamKeys() throws StorageException;
+
+  /**
+   * Attempt to delete the value corresponding to the given key. For some storage mediums we want to
+   * be able to abandon the delete under certain circumstances. For example, to abandon the attempt
+   * if we can't acquire a lock on the underlying storage.
+   *
+   * @param key The key to delete.
+   * @throws IncompleteOperationException the operation was incomplete and the key was not deleted
+   * @throws StorageException any other problem encountered during the deletion attempt.
    */
-  long removeAllKeysUnless(Predicate<byte[]> retainCondition) throws StorageException;
+  void tryDelete(byte[] key) throws IncompleteOperationException, StorageException;
```

Comment:
We should make this return `boolean` rather than throwing an exception since not completing the operation is a "normal" outcome, we shouldn't model it as an exception.